### PR TITLE
[camera] Update Android Camera Access Permission Error Codes

### PR DESCRIPTION
**Pull request by camsim99**
Addresses [Issue #101384](https://github.com/flutter/flutter/issues/101384) by updating error codes as per its proposal.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (camsim99) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:35]: @hellohuanlin I decided to keep this error because you can get it by instantiating the camera or requesting permissions twice at times too close together. However, the app would crash and other errors result, so I added the `return` below to avoid such.

Let me know what you think of the error code and message to make it most closely conform with the other platforms!

File: packages/camera/camera/CHANGELOG.md
Comment (hellohuanlin) [inline packages/camera/camera/CHANGELOG.md:6]: should this be a new version bump since we updated API?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (hellohuanlin) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:35]: Would it work if you just `return` but do not report this error dart side? 
What do you expect user to do when they have this error?

Comment (hellohuanlin) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:35]: On iOS side it's not possible because the requests happen on the main thread. So duplicate requests would queue up one after another.

File: packages/camera/camera/CHANGELOG.md
Comment (hellohuanlin) [inline packages/camera/camera/CHANGELOG.md:1]: Since we renamed the error code which is a breaking change, i think we should bump the major version. And since we are still in beta, major version is shifted to minor version, so it should be `0.10.0` here.

File: packages/camera/camera/CHANGELOG.md
Comment (hellohuanlin) [inline packages/camera/camera/CHANGELOG.md:3]: Updates *Android* camera access ...

**Review by hellohuanlin — APPROVED** [review]
Left a comment on the change log. Otherwise LGTM

File: packages/camera/camera/CHANGELOG.md
Comment (camsim99) [reply packages/camera/camera/CHANGELOG.md:1]: Ah gotcha, thanks!

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (blasten) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: how would a Dart user handle this case?

Comment (hellohuanlin) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: If I understand it correctly, this is a programming error that could be avoided by developers, which is different from other errors that depends on user interaction (`CameraAccessDenied`)

Comment (hellohuanlin) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: On iOS side it's not even possible because the requests happen on the main thread. So duplicate requests would queue up one after another.

Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: This change creates a codepath that doesn't call the result callback, which violates the engine API.

> If I understand it correctly, this error is a programming mistake that could be avoided by developers

First they have to know about it though; even developer mistakes should be sent back over the channel to Dart. They can be converted to some kind of `Error` objects (rather than `Exception` objects; see https://api.dart.dev/stable/2.16.2/dart-core/Error-class.html).

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:96]: It would be helpful to declare a set of string constants in a single location with a reference to the Dart code that handles these, so that it's obvious to maintainers (by looking at existing examples) that they shouldn't just be using random string values when adding new errors.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/camera/camera/CHANGELOG.md
Comment (stuartmorgan) [reply packages/camera/camera/CHANGELOG.md]: Are the iOS errors already using these names? It would be good to not make two separate breaking changes to `camera`, on for Android and one for iOS.

Comment (hellohuanlin) [reply packages/camera/camera/CHANGELOG.md]: The iOS didn't handle these error before, so it's not a breaking change. Android and web will be breaking changes.

Comment (stuartmorgan) [reply packages/camera/camera/CHANGELOG.md]: In that case it would be much better to hold this PR for the moment, make the `camera_web` breaking change, and then have this PR include bumping the major version of `camera_web`, so that this PR is a single breaking change that brings all platforms into alignment.

Breaking changes in app-facing plugin packages are disruptive to clients, so should be minimized when possible.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (hellohuanlin) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: Gotcha, good to learn. I think i'm fine with sending this to dart as an Error. 


Do we expect developers to catch these errors? Something like: 
```
try {
  controller.initialize()
} catch (CameraException e) {
  // handle exceptions like before 
} catch (CameraError e) {
  if (e.code == "DuplicateRequest") {
    print("This should not have happened. There is a bug in camera usage and need to fix it.")
  }
}
```

Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: No, the documentation for `Error` explicitly says developers should not catch them.

People may, but they shouldn't. The important thing is that we surface them as errors in the context of the Dart code, rather than silently swallowing serious programmer errors.

Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: ("silently" in the sense that an Android log message may never be noticed; an `Error` will be.)

Comment (hellohuanlin) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:37]: Gotcha. thanks for the explain.

File: packages/camera/camera/CHANGELOG.md
Comment (hellohuanlin) [reply packages/camera/camera/CHANGELOG.md]: The web change is not ready yet. How about a minor bump for Android to land this PR first, and then have the web change to do the major bump?

Comment (stuartmorgan) [reply packages/camera/camera/CHANGELOG.md]: You can't land a breaking change as a minor version bump; we follow semver.

Comment (camsim99) [reply packages/camera/camera/CHANGELOG.md]: @stuartmorgan Makes sense to me on holding this off. For now, though, made the changes if you can take a look! I wasn't sure how to reference Java code from Dart and vice versa, so if you have any feedback on that, I'd greatly appreciate it!

Comment (camsim99) [reply packages/camera/camera/CHANGELOG.md]: @hellohuanlin I renamed the error code, and since it wasn't in your design, if you can take a look, that'd be great!

File: packages/camera/camera/example/lib/main.dart
Comment (hellohuanlin) [inline packages/camera/camera/example/lib/main.dart:714]: I think you can remove this case since it's an error that we do not expect developers to handle, which is different from exceptions.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (camsim99) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:30]: @stuartmorgan Seems like to handle this error in `camera_platform_interface` I need to make a separate PR. Should that PR be landed before this one?

File: packages/camera/camera/CHANGELOG.md
Comment (blasten) [inline packages/camera/camera/CHANGELOG.md:3]: how is this a breaking change?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (blasten) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:111]: do we know if grantResults[1] always equals the audio?

File: packages/camera/camera/CHANGELOG.md
Comment (hellohuanlin) [reply packages/camera/camera/CHANGELOG.md:3]: @blasten because 0.3.0 is a breaking change on camera_web. We should probably provide some details here. Something like: 

```
**BREAKING CHANGE** Bumps default camera_web package version to 0.3.0 which renames permission exception code from `cameraPermission` to `CameraAccessDenied`. 
```

File: packages/camera/camera/CHANGELOG.md
Comment (blasten) [inline packages/camera/camera/CHANGELOG.md:4]: I think it would need to provide the steps to migrate from the previous version to the new version

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (camsim99) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:111]: Peeking at the implementation, I see that the order of `grantResults` mirrors the order in which permissions are requested, and we request them in that order.

File: packages/camera/camera/CHANGELOG.md
Comment (camsim99) [reply packages/camera/camera/CHANGELOG.md:4]: Can I ask for clarification here? Do you mean to note the new error code used?

Comment (hellohuanlin) [reply packages/camera/camera/CHANGELOG.md:4]: I think what @blasten meant (let me know if i misunderstood) is to add some migration instruction here, something like: 

"If you use handle the legacy `cameraPermission` exception code, please update it to use the new exception codes as described in the README. "

Comment (camsim99) [reply packages/camera/camera/CHANGELOG.md]: @blasten I updated it, but if you were referencing something else, just let me know!

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java
Comment (camsim99) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraPermissions.java:111]: Well, the [documentation](https://developer.android.com/reference/androidx/core/app/ActivityCompat.OnRequestPermissionsResultCallback#public-methods_1) does note that the results correspond to the list of permissions and another piece of verification is this sample app that they note in the source code as to how to use the method that uses the same method ([link](https://github.com/googlearchive/android-RuntimePermissions/blob/96612da3a0b4489fdf818a624a7e26fc768c65c9/Application/src/main/java/com/example/android/system/runtimepermissions/MainActivity.java#L254)).

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Pull request merged**
